src/System.Data.Common/src/System/Data/Common/DbProviderFactories.cs

```diff
@@ -0,0 +1,199 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+
+namespace System.Data.Common
+{
+    public static partial class DbProviderFactories
+    {
+        private struct ProviderRegistration
+        {
+            internal ProviderRegistration(string factoryTypeAssemblyQualifiedName, DbProviderFactory factoryInstance)
+            {
+                ADP.CheckArgumentLength(factoryTypeAssemblyQualifiedName, nameof(factoryTypeAssemblyQualifiedName));
+                this.FactoryTypeAssemblyQualifiedName = factoryTypeAssemblyQualifiedName;
+                this.FactoryInstance = factoryInstance;
+            }
+
+            internal string FactoryTypeAssemblyQualifiedName { get; }
+            /// <summary>
+            /// The cached instance of the type in <see cref="FactoryTypeAssemblyQualifiedName"/>. If null, this registation is seen as a deferred registration
+            /// and <see cref="FactoryTypeAssemblyQualifiedName"/> is checked the first time when this registration is requested through GetFactory().
+            /// </summary>
+            internal DbProviderFactory FactoryInstance { get; }
+        }
+
+        private static ConcurrentDictionary<string, ProviderRegistration> _registeredFactories = new ConcurrentDictionary<string, ProviderRegistration>();
+        private const string AssemblyQualifiedName = "AssemblyQualifiedName";
+        private const string InvariantName = "InvariantName";
+        private const string Name = "Name";
+        private const string Description = "Description";
+        private const string ProviderGroup = "DbProviderFactories";
```

Comment:
I would probably put something in the name to indicate which ones are column names, otherwise they could be defined as `nameof()` themselves, although that might look a bit weird.

Reply:
I'll rename the fields, good idea. I kept them together as it otherwise is a bit scattered. Now everything is discoverable in fixed locations (as one const indeed needs to be outside the method. )